### PR TITLE
bpo-24160: Fix breakpoints persistence across multiple pdb sessions

### PR DESCRIPTION
**Pull request by iritkatriel**
The information about breakpoints is split between Breakpoint.bplist/Breakpoint.bpbynumber (at class scope) and the breaks dictionary that lives on the Bdb() instance. The problems described in issue24160 occurs when those two parts get out of sync.

In interactive mode, each pdb.run() call creates a new instance of Bdb(). In this case the user would like the breakpoints to be remembered between calls. This PR adds a function _load_breaks to Bdb, which is called in its constructor and populates the instance's breaks dict with the data from the Breakpoint class.

Tests were added in both test_bdb (to test it directly) and in test_pdb (to test it with the interactive usecase).

https://bugs.python.org/issue24160

### CODE REVIEW

**Comment by pppery** [comment]
I would have gone the other way with this, and made the breakpoints for different Bdb instances completely independent, rather than completely dependent, but this works too.

**Comment by iritkatriel** [comment]
That was my first thought as well, but then I realised it would be annoying to have to set the same breakpoints again and again while debugging something.  (It’s easy to do ‘clear’ if that’s what you want).

File: Lib/bdb.py [hunk 1]
```diff
@@ -365,6 +367,12 @@ def set_quit(self):
     # Call self.get_*break*() to see the breakpoints or better
     # for bp in Breakpoint.bpbynumber: if bp: bp.bpprint().
 
+    def _add_to_breaks(self, filename, lineno):
+        """Add breakpoint to breaks, if not already there."""
+        list = self.breaks.setdefault(filename, [])
```
Comment (taleinat) [inline Lib/bdb.py]: We should use a more meaningful name here, and avoid shadowing the built-in `list`.

Comment (iritkatriel) [reply Lib/bdb.py]: oh! yeah

File: Lib/test/test_pdb.py [hunk 2]
```diff
@@ -213,6 +213,11 @@ def test_pdb_basic_commands():
     BAZ
     """
 
+def reset_Breakpoint_state():
+    import bdb
+    bdb.Breakpoint.next = 1
+    bdb.Breakpoint.bplist = {}
+    bdb.Breakpoint.bpbynumber = [None]
```
Comment (taleinat) [inline Lib/test/test_pdb.py:220]: I realize that this is copied from `test_bdb.py`, but perhaps:
```suggestion
        for bp in Breakpoint.bpbynumber:
        if bp:
            bp.deleteMe()
```

**Review by taleinat — COMMENTED** [review]
If possible, I think it would be god for the new tests to be slightly expanded to also include deleting breakpoints.

**Review by taleinat — COMMENTED** [review]
Perhaps now `Bdb.clear_all_breaks()` should use the new `clearBreakpoints` internally?

**Comment by iritkatriel** [comment]
> If possible, I think it would be god for the new tests to be slightly expanded to also include deleting breakpoints.

I don't see any tests for deleting breakpoint, and indeed coverage seems pretty grim:
https://codecov.io/gh/python/cpython/src/master/Lib/bdb.py

**Comment by iritkatriel** [comment]
> Perhaps now `Bdb.clear_all_breaks()` should use the new `clearBreakpoints` internally?

I think this could break something if some Bdb subclass changes the way this data is represented.

**Comment by iritkatriel** [comment]
It's messy because this class level state is accessed through self.

**Comment by taleinat** [comment]
> > Perhaps now `Bdb.clear_all_breaks()` should use the new `clearBreakpoints` internally?
> 
> I think this could break something if some Bdb subclass changes the way this data is represented.

True, let's leave it as is for now.

**Comment by taleinat** [comment]
Random question: How did "lineno" come to be shorthand for "line number"?

File: Lib/test/test_bdb.py [hunk 3]
```diff
@@ -951,6 +949,45 @@ def test_clear_at_no_bp(self):
         with TracerRun(self) as tracer:
             self.assertRaises(BdbError, tracer.runcall, tfunc_import)
 
+    def test_load_bps_from_previous_Bdb_instance(self):
```
Comment (taleinat) [inline Lib/test/test_bdb.py:952]: A few short comments and/or breaking this up into blocks could help with readability.

**Comment by taleinat** [comment]
@iritkatriel, this is looking good. I'll make a final review later and likely approve.

**Comment by iritkatriel** [comment]
> Random question: Maybe someone knows how "lineno" came to be shorthand for "line number"??

I asked google and it said something about Lenin.

**Comment by gvanrossum** [comment]
This looks like a feature, let's not backport it.

**Comment by iritkatriel** [comment]
Shall we try to get this into 3.10?

**Comment by gvanrossum** [comment]
SGTM. Is it ready? I haven't reviewed it yet.

**Comment by iritkatriel** [comment]
I've just rebased it. I think it is ready to merge. @taleinat had reviewed it a while back and we made some improvements.

**Review by gvanrossum — APPROVED** [review]
Okay, from a cursory inspection looks good. I trust @taleinat.

**Pull request merged**